### ISSUE TITLE
VTable Optimizations; Invalid application of 'sizeof' to an incomplete type llvm::pdb::PDBSymbolFunc / llvm::lto::InputFile

### ISSUE THREAD

**Issue by EmmaJaneBonestell** [llvm/llvm-project#55239]
Tested with clang 13, 14, 15. 
If attempting to compile LLVM with '-fforce-emit-vtables' and/or '-fwhole-program-vtables', the following files will fail with the respective errors:

<details>
  <summary>llvm/lib/DebugInfo/PDB/Native/LinePrinter.cpp</summary>

  ```
  In file included from /opt/venv/llvm- 
  project/llvm/lib/DebugInfo/PDB/Native/LinePrinter.cpp:9:
  In file included from /opt/venv/llvm- 
  project/llvm/include/llvm/DebugInfo/PDB/Native/LinePrinter.h:12:
  In file included from /opt/venv/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:14:
  In file included from /opt/venv/llvm- 
  project/llvm/include/llvm/ADT/SmallVector.h:28:
  In file included from /usr/bin/../lib/gcc/x86_64-linux- 
  gnu/11/../../../../include/c++/11/memory:76:
  /usr/bin/../lib/gcc/x86_64-linux- 
  gnu/11/../../../../include/c++/11/bits/unique_ptr.h:83:16: error: invalid application of 
  'sizeof' to an incomplete type 'llvm::pdb::PDBSymbolFunc'
          static_assert(sizeof(_Tp)>0,
                        ^~~~~~~~~~~
  /usr/bin/../lib/gcc/x86_64-linux- 
  gnu/11/../../../../include/c++/11/bits/unique_ptr.h:361:4: note: in instantiation of 
  member function 'std::default_delete<llvm::pdb::PDBSymbolFunc>::operator()' 
  requested here
            get_deleter()(std::move(__ptr));
            ^
  /usr/bin/../lib/gcc/x86_64-linux- 
  gnu/11/../../../../include/c++/11/bits/stl_construct.h:140:19: note: in instantiation 
  of member function 'std::unique_ptr<llvm::pdb::PDBSymbolFunc>::~unique_ptr' 
  requested here
        __pointer->~_Tp();
                    ^
  /usr/bin/../lib/gcc/x86_64-linux- 
  gnu/11/../../../../include/c++/11/bits/stl_construct.h:185:2: note: in instantiation of 
  function template specialization 
  'std::_Destroy_aux<false>::__destroy<std::unique_ptr<llvm::pdb::PDBSymbolFunc> *>' 
  requested here
          __destroy(__first, __last);
          ^
  /usr/bin/../lib/gcc/x86_64-linux- 
  gnu/11/../../../../include/c++/11/bits/alloc_traits.h:746:7: note: in instantiation of 
  function template specialization 
  'std::_Destroy<std::unique_ptr<llvm::pdb::PDBSymbolFunc> *>' requested here
        _Destroy(__first, __last);
        ^
  /usr/bin/../lib/gcc/x86_64-linux- 
  gnu/11/../../../../include/c++/11/bits/stl_vector.h:680:7: note: in instantiation of 
  function template specialization 
  'std::_Destroy<std::unique_ptr<llvm::pdb::PDBSymbolFunc> *, 
  std::unique_ptr<llvm::pdb::PDBSymbolFunc>>' requested here
          std::_Destroy(this->_M_impl._M_start, this->_M_impl._M_finish,
               ^
  /opt/venv/llvm-project/llvm/include/llvm/DebugInfo/PDB/UDTLayout.h:109:7: note: 
  in instantiation of member function 
  'std::vector<std::unique_ptr<llvm::pdb::PDBSymbolFunc>>::~vector' requested here
  class UDTLayoutBase : public LayoutItemBase {
        ^
  /opt/venv/llvm-project/llvm/include/llvm/DebugInfo/PDB/PDBTypes.h:41:7: note: 
  forward declaration of 'llvm::pdb::PDBSymbolFunc'
  class PDBSymbolFunc;
        ^
  1 error generated

  ```
</details>

<details>

  <summary>tools/lld/MachO/CMakeFiles/lldMachO.dir/...</summary>

  Arch/ARM64_32.cpp 
  Arch/ARM64Common.cpp 
  Arch/ARM64.cpp 
  Arch/ARM.cpp 
  Arch/X86_64.cpp 
  ConcatOutputSection.cpp 
  Dwarf.cpp 
  ExportTrie.cpp 
  ICF.cpp 
  InputSection.cpp 
  MapFile.cpp 
  MarkLive.cpp 
  ObjC.cpp 
  OutputSection.cpp 
  OutputSegment.cpp 
  Relocations.cpp 
  SectionPriorities.cpp 
  Symbols.cpp 
  SymbolTable.cpp 
  SyntheticSections.cpp 
  UnwindInfoSection.cpp 
  Writer.cpp

  ```
  In file included from /opt/venv/llvm-project/lld/MachO/Arch/ARM64Common.cpp:9:
  In file included from /opt/venv/llvm-project/lld/MachO/Arch/ARM64Common.h:12:
  In file included from /opt/venv/llvm-project/lld/MachO/InputFiles.h:13:
  In file included from /opt/venv/llvm-project/lld/MachO/Target.h:13:
  In file included from /opt/venv/llvm-project/lld/MachO/Relocations.h:13:
  In file included from /opt/venv/llvm- 
  project/llvm/include/llvm/ADT/PointerUnion.h:20:
  In file included from /opt/venv/llvm-project/llvm/include/llvm/ADT/STLExtras.h:38:
  In file included from /usr/bin/../lib/gcc/x86_64-linux- 
  gnu/11/../../../../include/c++/11/memory:76:
  /usr/bin/../lib/gcc/x86_64-linux- 
  gnu/11/../../../../include/c++/11/bits/unique_ptr.h:83:16: error: invalid application of 
  'sizeof' to an incomplete type 'llvm::lto::InputFile'
          static_assert(sizeof(_Tp)>0,
                        ^~~~~~~~~~~
  /usr/bin/../lib/gcc/x86_64-linux- 
  gnu/11/../../../../include/c++/11/bits/unique_ptr.h:361:4: note: in instantiation of 
  member function 'std::default_delete<llvm::lto::InputFile>::operator()' requested here
            get_deleter()(std::move(__ptr));
            ^
  /opt/venv/llvm-project/lld/MachO/InputFiles.h:247:7: note: in instantiation of 
  member function 'std::unique_ptr<llvm::lto::InputFile>::~unique_ptr' requested here
  class BitcodeFile final : public InputFile {
        ^
  /opt/venv/llvm-project/lld/MachO/InputFiles.h:30:7: note: forward declaration of 
  'llvm::lto::InputFile'
  class InputFile;
        ^
  1 error generated
  ```

</details>

**Comment by fhahn** [comment]
Would it be possible to attach the preprocessed file that is causing the error?

**Comment by EmmaJaneBonestell** [comment]
@fhahn  Apologies for the delay. If you need anything else, I should be much more responsive now. 

Here are outputs from preprocessing/-gen-reproducer:

https://drive.google.com/drive/folders/1Nqs8l8eb4h_0Q8mTk_RMSQPEgJLY8Dfz?usp=sharing

**Comment by EmmaJaneBonestell** [comment]
I'm not sure if it meets any style or other guidelines, since I had to fix ambiguous references after including the missing header, but this patch should fix the issue.

https://gist.githubusercontent.com/EmmaJaneBonestell/6bee1ef3ed9217fa99ffa659ece7bdfb/raw/fd4e8c7550bc4c7cab876f10933737700c82e312/0001-Fix-incomplete-type-errors-revealed-when-virtual-fun.patch

**Comment by JohelEGP** [comment]
May be related to #58206.
Are you compiling in C++23 mode?

**Comment by EmmaJaneBonestell** [comment]
> May be related to #58206. Are you compiling in C++23 mode?

I was not compiling in C++23 mode.  As far as I can tell, it was and still is a forward declaration error, and I already provided a patch for it in this thread. It's always fun to make detailed bug reports, then have to find and fix the issue yourself, provide that patch, and then go ignored. 🤷🏼‍♀️

**Comment by JohelEGP** [comment]
Patches meant to be contributed go to <https://reviews.llvm.org/>.
Although I don't see that at <https://clang.llvm.org/get_involved.html>,
so I'm not sure exactly how to get there.

**Comment by EmmaJaneBonestell** [comment]
I'll send it there at some point then. I just followed the llvm bug reporting guide, which doesn't mention that whatsoever. Not exactly an effective method of fixing bugs when one's own documentation and personnel don't merge the gap from bug report to a fix. From the silence, I just presumed my patch was unacceptable, and no one cared to fix the problem.

Thank you for pointing me in the right direction.

**Comment by dwblaikie** [comment]
I might've missed it, but can someone explain this issue in more detail? Is the code correct and the feature (`-fforce-emit-vtables) requires code to be written in a particular way? (if so, it'd be good to document that restriction and see if the LLVM project is willing to accept such a restriction)
Is it that this feature identifies invalid code that's not been diagnosed previously? (if so, it'd be good to have an explanation of what kind of invalid code is silently accepted usually, but identified by this feature)

**Comment by EmmaJaneBonestell** [comment]
@dwblaikie No. Optimizations related to `-fforce-emit-vtables` revealed a forward-declaration bug in llvm's source code and no one else has bothered to fix it. I haven't bothered submitting a fix to reviews.llvm.org because they make it a pain in the ass to submit even the smallest of fixes for review. It's ironic that clang cannot build llvm with some of its supposedly stable/non-experimental optimization flags.

**Comment by dwblaikie** [comment]
Could you provide a representative example of the invalid code that's silently accepted without these flags, but fails to compile with these flags? (That's equivalent to the situation in the llvm source code itself)

I'd like to better understand the root cause here.

**Comment by EmmaJaneBonestell** [comment]
@dwblaikie Is there a particular reason you are interested in this?

 If you look at the patch file I provided, you can easily see what the problem was/is, but only if you understand what devirtualization is and how that is accomplishable at compile time rather than runtime. 

It requires the entire class hierarchy to be completely visible at compile time. As it tries to make these optimizations, it fails. If you note in the compiler error output I provided for both sets of files, the only thing that fails is the sizeof() function because it cannot work on forward declarations. Normally those calls would have been devirtualized at runtime, when it would then have access to that information.

**Comment by JohelEGP** [comment]
> I haven't bothered submitting my fix to reviews.llvm.org because they make it a pain in the ass to submit even the smallest of fixes for review.

Is this particular to the Phabricator?
Now they're using GitHub's PRs.

**Comment by EmmaJaneBonestell** [comment]
> > I haven't bothered submitting my fix to reviews.llvm.org because they make it a pain in the ass to submit even the smallest of fixes for review.
> 
> Is this particular to the Phabricator? Now they're using GitHub's PRs.

Yep. Hadn't realized it had switched. I just got email notifications about the replies here. Thanks for the heads up.

**Comment by dwblaikie** [comment]
> @dwblaikie Is there a particular reason you are interested in this?

Reviewing the proposed change https://github.com/llvm/llvm-project/pull/65456 and wanted to discuss what the underlying issue is/what the suitable solution is.

> If you look at the patch file I provided, you can easily see what the problem was/is, but only if you understand what devirtualization is and how that is accomplishable at compile time rather than runtime.

I had trouble understanding what circumstances were needed to reproduce the issue in isolation. But playing around with it for a while I managed to get to this:

https://godbolt.org/z/EYn4P8rnY

Which shows a case where it compiles without the flag, but fails with the flag.

I don't actually know if this code is valid or invalid, as-is. I think with an explicit instantiation like this ( https://godbolt.org/z/YEePG1dEK ) the code probably is valid, and rejecting that code is probably incorrect (but maybe a necessary tradeoff/thing that people using `-fforce-emit-vtables` are willing to accept)

I'm not sure it's worth supporting this in LLVM - but the additional includes in the proposed patch probably aren't a huge deal - I just worry about the longer term support issues of this mode.

**Comment by ElDavoo** [comment]
llvm 19.1.3 does compile with `-fwhole-program-vtables`, but not with `-fforce-emit-vtables`
